### PR TITLE
[Do NOT Squash] Merge retiarii dev branch to master

### PR DESCRIPTION
**Pull request by liuzhe-lz**

### CODE REVIEW

**Comment by liuzhe-lz** [comment]
Must review:
@liuzhe-lz @QuanluZhang @ultmaster @qfyin

File: nni/retiarii/graph.py [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+        for graph in self.graphs.values():
+            nodes = graph.get_nodes_by_label(label)
+            matched_nodes.extend(nodes)
+        return matched_nodes
+
+    def get_by_name(self, name: str) -> Union['Graph', 'Node']:
+        """
+        Find the graph or node that have the given name space name.
+        """
+
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:176]: Huh?

File: nni/retiarii/utils.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    def __init__(self):
+        self.recorded_arguments = None
+    
+    def __enter__(self):
+        enable_record_args()
+        return self
+
+    def __exit__(self, exc_type, exc_value, tb):
+        if exc_type is not None:
+            traceback.print_exception(exc_type, exc_value, tb)
```
Comment (liuzhe-lz) [inline nni/retiarii/utils.py:21]: What's the expected behavior when exception raised?

File: nni/retiarii/utils.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            # return False # uncomment to pass exception through
+        self.recorded_arguments = get_records()
+        disable_record_args()
```
Comment (liuzhe-lz) [inline nni/retiarii/utils.py:24]: So this can't be nested?

File: nni/retiarii/trial_entry.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+Assuming execution engine is BaseExecutionEngine.
+"""
+import os
+
+from .execution.base import BaseExecutionEngine
+from .execution.cgo_engine import CGOExecutionEngine
+
+if __name__ == '__main__':
+    if os.environ.get('CGO') == 'true':
```
Comment (liuzhe-lz) [inline nni/retiarii/trial_entry.py:12]: Where is this set?

File: nni/retiarii/trainer/interface.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    Currently, our trainer = LightningModule + LightningTrainer. We might want to separate these two things
+    in future.
+
+    Trainer has a ``fit`` function with no return value. Intermediate results and final results should be
+    directly sent via ``nni.report_intermediate_result()`` and ``nni.report_final_result()`` functions.
+    """
+    def __init__(self, *args, **kwargs):
+        module = self.__class__.__module__
+        if module is None or module == str.__class__.__module__:
```
Comment (liuzhe-lz) [inline nni/retiarii/trainer/interface.py:24]: Please fix this.

File: nni/retiarii/nn/pytorch/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+from .nn import *
```
Comment (liuzhe-lz) [inline nni/retiarii/nn/pytorch/__init__.py:1]: `import *` should always be used together with `__all__`.

File: nni/retiarii/integration.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+import nni
+from nni.runtime.msg_dispatcher_base import MsgDispatcherBase
+from nni.runtime.protocol import send, CommandType
+from nni.utils import MetricType
+
+from . import utils
+from .graph import MetricData
+
+
+_logger = logging.getLogger('nni.msg_dispatcher_base')
```
Comment (liuzhe-lz) [inline nni/retiarii/integration.py:15]: Why this logger?

File: nni/retiarii/graph.py [hunk 8]
```diff
@@ -0,0 +1,654 @@
...
+        }
+
+
+class IllegalGraphError(ValueError):
+    def __init__(self, graph, *args):
+        self._debug_dump_graph(graph)
+        super().__init__(*args)
+
+    @staticmethod
+    def _debug_dump_graph(graph):
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:650]: Why is this a static method of error class?

File: nni/retiarii/graph.py [hunk 9]
```diff
@@ -0,0 +1,654 @@
...
+        self._root_graph_name: str = '_model'
+        self.graphs: Dict[str, Graph] = {}
+        self.training_config: TrainingConfig = TrainingConfig('foo', {})
+
+        self.history: List[Model] = []
+
+        self.metric: Optional[MetricData] = None
+        self.intermediate_metrics: List[MetricData] = []
+
+        self._last_uid: int = 0  # FIXME: this should be global, not model-wise
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:109]: Need to fix. (my bad)

File: nni/retiarii/mutator.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        self._cur_choice_idx: Optional[int] = None
+
+    def bind_sampler(self, sampler: Sampler) -> 'Mutator':
+        """
+        Set the sampler which will handle `Mutator.choice` calls.
+        """
+        self.sampler = sampler
+        return self
+
+    def apply(self, model: Model) -> Model:
```
Comment (liuzhe-lz) [inline nni/retiarii/mutator.py:50]: Can anyone suggest a better name? This name looks like "mutate in-place".

File: nni/retiarii/mutator.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        Dry run mutator on a model to collect choice candidates.
+        If you invoke this method multiple times on same or different models,
+        it may or may not return identical results, depending on how the subclass implements `Mutator.mutate()`.
+        """
+        sampler_backup = self.sampler
+        recorder = _RecorderSampler()
+        self.sampler = recorder
+        new_model = self.apply(model)
+        self.sampler = sampler_backup
+        return recorder.recorded_candidates, new_model
```
Comment (liuzhe-lz) [inline nni/retiarii/mutator.py:78]: As an user API, I think the return type is too complicated.
I suggest to make this method "static" and accept multiple mutators:
```python
class Mutator:
    @staticmethod  # optionally
    def dry_run(mutator: Union[Mutator, Iterable[Mutator]], model: Model) -> list[list[Choice]]:
        ...

Mutator.dry_run([mutator1, mutator2, mutator3], model)
Mutator.dry_run(single_mutator, model)
single_mutator.dry_run(model)  # if remove "@staticmethod" decorator above
```

File: nni/retiarii/graph.py [hunk 12]
```diff
@@ -0,0 +1,654 @@
...
+
+MetricData = Any
+"""
+Graph metrics like loss, accuracy, etc.
+
+# Maybe we can assume this is a single float number for first iteration.
+"""
+
+
+class TrainingConfig:
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:24]: Please write real doc for this class.
I copied it from artifact code and have no idea how it works.

File: nni/retiarii/graph.py [hunk 13]
```diff
@@ -0,0 +1,654 @@
...
+from .operation import Cell, Operation, _IOPseudoOperation
+
+__all__ = ['Model', 'ModelStatus', 'Graph', 'Node', 'Edge', 'IllegalGraphError', 'MetricData']
+
+
+MetricData = Any
+"""
+Graph metrics like loss, accuracy, etc.
+
+# Maybe we can assume this is a single float number for first iteration.
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:20]: What's the real format? Float or dict?

File: nni/retiarii/graph.py [hunk 14]
```diff
@@ -0,0 +1,654 @@
...
+    @property
+    def outgoing_edges(self) -> List['Edge']:
+        return [edge for edge in self.graph.edges if edge.head is self]
+
+    @property
+    def cell(self) -> Graph:
+        assert isinstance(self.operation, Cell)
+        return self.graph.model.graphs[self.operation.parameters['cell']]
+
+    def update_label(self, label: str) -> None:
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:513]: What is label???

File: nni/retiarii/graph.py [hunk 15]
```diff
@@ -0,0 +1,654 @@
...
+        return f'Graph(id={self.id}, name={self.name}, ' + \
+            f'input_names={self.input_node.operation.io_names}, ' + \
+            f'output_names={self.output_node.operation.io_names}, ' + \
+            f'num_hidden_nodes={len(self.hidden_nodes)}, num_edges={len(self.edges)})'
+
+    @property
+    def nodes(self) -> List['Node']:
+        return [self.input_node, self.output_node] + self.hidden_nodes
+
+    def _add_input(self, input_name) -> None:
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:257]: This is for PyTorch converter only. Suggest move into converter package.

File: nni/retiarii/graph.py [hunk 16]
```diff
@@ -0,0 +1,654 @@
...
+        self.add_edge((new_node, None), (edge.tail, edge.tail_slot))
+        self.del_edge(edge)
+        return new_node
+
+    # mutation
+    def add_edge(self, head: Tuple['Node', Optional[int]], tail: Tuple['Node', Optional[int]]) -> 'Edge':
+        assert head[0].graph is self and tail[0].graph is self
+        return Edge(head, tail, _internal=True)._register()
+
+    def del_edge(self, edge: 'Edge') -> None:
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:303]: Please sync the name. Either rename this to `remove_edge`, or rename `Edge.remove` to `del`.

File: nni/retiarii/graph.py [hunk 17]
```diff
@@ -0,0 +1,654 @@
...
+            op = Operation.new(operation_or_type, parameters, name)
+        new_node = Node(self, self.model._uid(), name, op, _internal=True)._register()
+        # update edges
+        self.add_edge((edge.head, edge.head_slot), (new_node, None))
+        self.add_edge((new_node, None), (edge.tail, edge.tail_slot))
+        self.del_edge(edge)
+        return new_node
+
+    # mutation
+    def add_edge(self, head: Tuple['Node', Optional[int]], tail: Tuple['Node', Optional[int]]) -> 'Edge':
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:299]: The type hint is too confusing. I'll deal with it.

File: nni/retiarii/graph.py [hunk 18]
```diff
@@ -0,0 +1,654 @@
...
+        """
+        Fork the model and returns corresponding graph in new model.
+        This shortcut might be helpful because many algorithms only cares about "stem" subgraph instead of whole model.
+        """
+        return self.model.fork().graphs[self.name]
+
+    def __eq__(self, other: object) -> bool:
+        return self is other
+
+    def _fork_to(self, model: Model, name_prefix='') -> 'Graph':
```
Comment (liuzhe-lz) [inline nni/retiarii/graph.py:370]: Where is "name_prefix" used?

File: nni/retiarii/nn/pytorch/nn.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
+import inspect
+import logging
+import torch
+import torch.nn as nn
+from typing import (Any, Tuple, List, Optional)
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.INFO)
```
Comment (liuzhe-lz) [inline nni/retiarii/nn/pytorch/nn.py:8]: Why set level here?

File: nni/retiarii/execution/base.py [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+        for listener in self._listeners:
+            _logger.warning('resources: {}'.format(listener.resources))
+            if not listener.has_available_resource():
+                _logger.warning('There is no available resource, but trial is submitted.')
+            listener.on_resource_used(1)
+            _logger.warning('on_resource_used: {}'.format(listener.resources))
+
+    def _request_trial_jobs_callback(self, num_trials: int) -> None:
+        for listener in self._listeners:
+            listener.on_resource_available(1 * num_trials)
```
Comment (liuzhe-lz) [inline nni/retiarii/execution/base.py:76]: `1 * `?

File: test/ut/retiarii/advisor_entry.py [hunk 21]
```diff
@@ -0,0 +1,4 @@
+import os
+import sys
+
+from nni.retiarii.integration import RetiariiAdvisor
```
Comment (QuanluZhang) [inline test/ut/retiarii/advisor_entry.py:4]: is this file still necessary?

**Review by QuanluZhang — APPROVED** [review]

**Review by ultmaster — APPROVED** [review]

File: nni/retiarii/integration.py (hunk 7 above)
Comment (hzhua) [reply nni/retiarii/integration.py:12]: It should be removed. I added it for debug.

Comment (ultmaster) [reply nni/retiarii/integration.py:12]: To print message to dispatcher log? Any suggestions?

File: nni/retiarii/trial_entry.py (hunk 4 above)
Comment (hzhua) [reply nni/retiarii/trial_entry.py:12]: Which set? If user sets "CGO=true" in the environment variables, it uses CGOExecutionEngine.

File: test/ut/retiarii/advisor_entry.py (hunk 21 above)
Comment (ultmaster) [reply test/ut/retiarii/advisor_entry.py:4]: We don't need it any more I think.

File: nni/retiarii/trial_entry.py (hunk 4 above)
Comment (liuzhe-lz) [reply nni/retiarii/trial_entry.py:12]: There are training services other than "local", where trials will run on ssh server or cloud. So environment variables cannot be set by user.

Comment (QuanluZhang) [reply nni/retiarii/trial_entry.py:12]: yes, this part does not work for other training services. we should refactor this part.
currently, retiarii only supports local training service...

Comment (hzhua) [reply nni/retiarii/trial_entry.py:12]: We can add "CGO=true" in the trial_command of RetiariiExeConfig if cross-graph optimization is enabled.
